src/slg/samplers/sobolsequence.cpp

```diff
@@ -44,13 +44,19 @@ void SobolSequence::RequestSamples(const u_int size) {
 }
 
 u_int SobolSequence::SobolDimension(const u_int index, const u_int dimension) const {
-	const u_int offset = dimension * SOBOL_BITS;
+	u_int offset = dimension * SOBOL_BITS;
 	u_int result = 0;
 	u_int i = index;
 
-	for (u_int j = 0; i; i >>= 1, j++) {
-		if (i & 1)
-			result ^= directions[offset + j];
+	while (i) {
+		result ^= ((i & 1) * directions[offset]);
+		i >>= 1; offset++;
+		result ^= ((i & 1) * directions[offset]);
+		i >>= 1; offset++;
+		result ^= ((i & 1) * directions[offset]);
+		i >>= 1; offset++;
+		result ^= ((i & 1) * directions[offset]);
+		i >>= 1; offset++;
```

Comment:
There is no longer a bounds check on `i` for 3 of the 4 indexes into `directions` before the loop repeats, how do you ensure the `offset` is always a valid index?

Reply:
Before the change and after the change `i` wasn't used as an index into `directions`. Whenever the loop was hitting a `1` bit in `i` it would index into `directions` by that bit's offset. The way `directions` defined, starting from the `offset` it has `SOBOL_BITS` range for access, so in the new code the access to `directions` is valid and when `i` becomes `0` none of these `directions` values will be used in the result because of `&` with `(0 & 1)`.